### PR TITLE
Fix arguments and script name to match interface

### PR DESCRIPTION
**Pull request by AurelienFT**
Hello,

This is my first contribution to the Exercism project. I saw on the track of Rust that the representer wasn't working so I read the documentation of a representer and I want to contribute to improve the Rust one.

When I started to look at the repository I saw that this repository don't match the [interface doc](https://exercism.org/docs/building/tooling/representers/interface) and so in this PR I try to make this repository compatible with the doc.

I use this PR to get involved in the project slowly but in the future I would like to improve a lot the representer

I ask @senekor for a review because I saw that you were the only one last active on this repository but sorry if I didn't tag the correct person and if you can redirect me it would be nice :)

### CODE REVIEW

**Comment by senekor** [comment]
Hi @AurelienFT 

Please, in the future ask on the forum first before starting work on contributions, as the bot explains. I think it's awesome that you wanna help out though.

I don't have time to do an in-depth review right now, but this looks fine from reading the diff and the docs you linked. @ErikSchierboom needs to sign off on this, the tooling repos are owned by maintainers-admin.

**Comment by AurelienFT** [comment]
No problems I have made a forum post after seeing the bot response. Ty for kind review and waiting for the review :)

File: Dockerfile
Comment (AurelienFT) [inline Dockerfile:20]: Note for the review : I don't know why we target "x86_64-unknown-linux-musl" especially don't wanted to change that because I didn't know why it was there but if there is explanation I would be happy to know them !

Comment (senekor) [reply Dockerfile:20]: I assume you're wondering about `linux-gnu` versus `linux-musl`. `linux-gnu` is more common, it relies on `glibc`, a popular implementation of the C standard library by GNU. The Alpine docker image we're using (because it is a very small base) uses `musl` as its standard library instead of `glibc`. Let me know if there's more you'd like me to explain.

Comment (AurelienFT) [reply Dockerfile:20]: Yes I wasn't able to comment in upper lines, you assumed well 😆 . Ok can 't test with the Docker locally as i'm on M1 Apple I will try to make an other docker locally for me to test in the next days.

**Comment by AurelienFT** [comment]
I have been told that the PR will be review by Wednesday. I took the time to add everything to make the representer fully compatible and, even if it doesn't catch all the normalizations possible (should be in a second time from my point of view), output the correct format and could be considered as a v1 with the correct base to improve.

File: Dockerfile
Comment (senekor) [reply Dockerfile:20]: I see. That's probably an issue with `x86_64` though, the CPU architecture. I guess it should work the same way if you change it to test locally. But you can't commit that, because our servers are definitely x86 based. You could try with `aarch64-unknown-linux-musl`.

File: src/lib.rs
Comment (senekor) [inline src/lib.rs:19]: Looking at the docs, I think this is supposed to be called `representation.txt`. Not sure if the file extension is important for it to be picked up.

File: Dockerfile
Comment (iHiD) [reply Dockerfile:20]: FYI. I believe you should be able to run x86 images on your mac in general, using the `--platform linux/amd64` flag. See https://forums.docker.com/t/run-x86-intel-and-arm-based-images-on-apple-silicon-m1-macs/117123 for example.

On my machine (M2) I run the normal exercism x86 docker images. These are my docker settings in case this helps:

Screenshot 2023-09-10 at 00 16 56

File: src/lib.rs
Comment (AurelienFT) [reply src/lib.rs:19]: Yes but the run.sh that is ran by the docker make a fmt on the .rs before converting it to .txt. I think to have a fully Rust solution we could run the fmt in the rust code directly but it might be too much for this PR

File: Dockerfile
Comment (AurelienFT) [reply Dockerfile]: Thanks for the solution, worked well. I fixed the docker that wasn't working then and it simplify it a bit. Closing this thread as the main purpose of it has been answered

File: src/lib.rs
Comment (senekor) [reply src/lib.rs:19]: Got it, thanks! It's fine the way it is.

File: Dockerfile
Comment (senekor) [inline Dockerfile:17]: Deleting this is a problem. This line creates a new... layer? stage? I'm not sure exactly what it's called. But essentially, everthing after it creates an image separate from what came before it.

The reason we do this here is we don't want to keep everthing around from the build in terms of storage space. All the tooling containers combined need some amount of storage space, which every server needs to have at least. So keeping the size of these images small contributes to lower server costs for exercism.

Judging from your commit message, the reason you did this is because the container didn't find `rustfmt`. In that case, we need to find a way to install rustfmt in the alpine image, without wasting space.

**Review by senekor — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (senekor) [reply Dockerfile:17]: I don't have time to investigate the best way to do that. The releases page of `rustfmt` doesn't have prebuilt binaries for our target.

Two options I can think of right now:

Clone the [rustfmt repo](https://github.com/rust-lang/rustfmt) in the build stage, compile from source for our target, copy the binary over to the alpine stage

Maybe `rustup` can be coerced into downloading tooling for a different target than the one of the current machine? Then just copy it over. That would probably be the fastest way. But I just investigated this, it looks like rustup installes its components as dynamically linked executables, I don't think we want to mess with that.

Comment (senekor) [reply Dockerfile:17]: The thing is this, installing the entire Rust toolchain just to have `rustfmt` is way too much. One toolchain is about 500 MiB, and I personally don't know the best way to strip that down to only what we need.

Comment (senekor) [reply Dockerfile:17]: Building from source might be the best option. To save ourselves the trouble of cross-compiling, we could change the build stage base from debian to alpine as well :thinking:

Comment (AurelienFT) [reply Dockerfile:17]: Ok I will try to readd this step today. Thanks for taking time to search on it !

Comment (AurelienFT) [reply Dockerfile:17]: What about running rustfmt on the rust binary directly using it as a dependency to avoid having to install it like this code : https://github.com/bcmyers/num-format/blob/v0.1.2/num-format-dev/src/rustfmt.rs

Comment (senekor) [reply Dockerfile:17]: Yes, that would be a very clean solution. It also entails compiling `rustfmt` from source, so we still incur the additional build time. But both `run.sh` and the dockerfile can be simplified greatly with this approach.

Not all programs that are primarily meant to be run as standalone executables have good support for being embedded as a library. I'm optimistic, but it must be investigated.

Comment (AurelienFT) [reply Dockerfile:17]: Ok I will try today ! Thanks for taking time to answer me on this PR :)

Comment (AurelienFT) [reply Dockerfile:17]: As you expected it's quite hard to run rustfmt as a library (in stable) but I found this project made by the syn maintainer (used by syn internally) : https://github.com/dtolnay/prettyplease . It seems to be a perfect fit for us as we are managing syn AST.

What do you think about removing the first and second call to rustfmt to only one call in the rust code to this library only AFTER the code has been normalized ?

Comment (senekor) [reply Dockerfile:17]: Yes that sounds great!

Comment (AurelienFT) [reply Dockerfile]: It works well now I have the same dockerfile as before and added the [`run-in-docker.sh`](https://github.com/exercism/python-representer/blob/main/bin/run-in-docker.sh) from `python-representer` to make local testing using docker more easy.

**Comment by AurelienFT** [comment]
Following [this discussion ](https://github.com/exercism/rust-representer/pull/29#discussion_r1320693137) we integrated the formatting directly in the rust code instead of a post-normalize step in a script which cause the tests to be reformatted to test the correct format.

**Review by senekor — APPROVED** [review]
This is fantastic work, thank you!

File: bin/run-in-docker.sh
Comment (ErikSchierboom) [inline bin/run-in-docker.sh:29]: In general we don't specify the platform. I think it'll work in production, but might be worth checking.

File: bin/run-in-docker.sh
Comment (ErikSchierboom) [inline bin/run-in-docker.sh:17]: This is probably copied from the analyzer? Could you update it to the representer? Feel free to copy from https://github.com/exercism/generic-representer/tree/main/bin

File: bin/run-in-docker.sh
Comment (ErikSchierboom) [reply bin/run-in-docker.sh:29]: You could use this as the image name:

```suggestion
docker build --platform linux/amd64 --rm --no-cache -t exercism/rust-representer .
```

With that convention, we can do some caching in GHA

File: bin/run-in-docker.sh
Comment (ErikSchierboom) [inline bin/run-in-docker.sh:39]: This command is missing some of the arguments that make it match our production environment:

```suggestion
docker run \
    --rm \
    --network none \
    --read-only \
    --mount type=bind,source="${PWD}/${2}",destination=/solution \
    --mount type=bind,source="${PWD}/${output_dir}",destination=/output \
    --mount type=tmpfs,destination=/tmp \
    exercism/rust-representer $1 /solution/ /output/
```

**Review by ErikSchierboom — COMMENTED** [review]
Couple of minor comments.

Once this PR is merged, I would recommend creating some golden tests that run in Docker. I'll help put up some scaffolding for that once this is merged.

**Comment by AurelienFT** [comment]
> Once this PR is merged, I would recommend creating some golden tests that run in Docker.

If it needs to follow the generic-representer you linked it shouldn't be much change and I can try do it except if you have time to do it and want to do it because I think you will be more faster than me xD

**Comment by ErikSchierboom** [comment]
> If it needs to follow the generic-representer you linked it shouldn't be much change and I can try do it except if you have time to do it and want to do it because I think you will be more faster than me xD

Be my guest!

**Comment by ErikSchierboom** [comment]
I've just done a small test and the change works nicely. I did find one bug that I think should be fixed before we merge this

This code:

```rust
pub fn is_leap_year(year: u64) -> bool {
    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
}   
```

is converted to:

```
pub fn PLACEHOLDER_1(PLACEHOLDER_2: u64) -> bool {
    PLACEHOLDER_2 % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
}
```

You can see that the placeholder is only changed once, there are still two `year` identifiers that should be replaced. If this could be fixed before this PR is merged, we won't have to re-run any representations.

**Comment by AurelienFT** [comment]
Ok I will fix it. I have a question about this : 

> If this could be fixed before this PR is merged, we won't have to re-run any representations.

We can't re-run representers runs in the future ? Because I think there is a lot of things that could be improved to normalize even more but on this PR I just wanted to have a first version.

Imagine we provide an improvement in the coming weeks then we can't re-run the representer on already normalized solutions to improve the previous result ?

**Comment by AurelienFT** [comment]
@ErikSchierboom The leap year test now works and has been added to the test suite of this repository

**Comment by ErikSchierboom** [comment]
> We can't re-run representers runs in the future ? Because I think there is a lot of things that could be improved to normalize even more but on this PR I just wanted to have a first version.

We can, but it is quite a strain on our infrastructure, so when we can prevent it, it is a good idea.

**Comment by ErikSchierboom** [comment]
> We can't re-run representers runs in the future ? Because I think there is a lot of things that could be improved to normalize even more but on this PR I just wanted to have a first version.
> 
> Imagine we provide an improvement in the coming weeks then we can't re-run the representer on already normalized solutions to improve the previous result ?

If you expect to fix/improve the representer in the coming weeks, it could be useful to temporarily disable the representer as it wasn't working anyway, and then re-enable once those changes have been applied.

**Comment by AurelienFT** [comment]
> If you expect to fix/improve the representer in the coming weeks, it could be useful to temporarily disable the representer as it wasn't working anyway, and then re-enable once those changes have been applied.

Yes I think we could do much much better, I wanted to expose my idea in a forum post and then code them here. I'm not sure if I will have the time. Can we say that we merge this but don't launch it for now live and I have like 2 weeks to make the improvements and if you don't hear me back after 2 weeks you launch it as it is ?

**Review by ErikSchierboom — APPROVED** [review]

**Comment by ErikSchierboom** [comment]
We've had some discussions internally and concluded that it's fine to merge it. Thanks for the hard work!

**Pull request merged**
